### PR TITLE
Allow empty strings, instead of treating them as missing

### PR DESCRIPTION
**Pull request by wez**
Note that this PR is based on https://github.com/instant-labs/instant-xml/pull/61, but doesn't depend upon it; I didn't want to just push this straight into that PR, and I didn't want to create too many more branches to reconcile the version of this crate that is pulled into my downstream project!

### CODE REVIEW

**Comment by wez** [comment]
once both this and #61 are merged, I think it's reasonable to cut a new release of instant-xml; I've been sitting on both of these changes for the past week or so and haven't run into other weird cases around them in that time.

**Comment by wez** [comment]
Hmm, hold on merging this: I tweaked the implementation and tests to make the PR look nicer, but running the tests in my downstream project now fail. I'm investigating!

**Comment by wez** [comment]
OK, the problem case was with an empty nested direct field being treated as missing:

```rust
#[derive(Debug, PartialEq, FromXml)]
struct ArtUri {
    #[xml(direct)]
    uri: String,
}

#[derive(Debug, PartialEq, FromXml)]
struct Container {
    art: Option<ArtUri>,
}

#[test]
fn container_empty_string() {
    assert_eq!(
        from_str("<Container><ArtUri></ArtUri></Container>"),
        Ok(Container {
            art: Some(ArtUri {
                uri: "".to_string()
            })
        })
    );
}
```

 I've adjusted the codegen for that case; it now looks like this; I've annotated the new sections with `HERE`:

```rust
    fn deserialize<'cx>(
        into: &mut Self::Accumulator,
        field: &'static str,
        deserializer: &mut ::instant_xml::Deserializer<'cx, 'xml>,
    ) -> ::std::result::Result<(), ::instant_xml::Error> {
        use ::instant_xml::de::Node;
        use ::instant_xml::{Accumulate, Error, FromXml, Id, Kind};
        enum __Elements {
            __Ignore,
        }
        enum __Attributes {
            __Ignore,
        }
        let mut __value0 = <String as FromXml>::Accumulator::default();
        let mut seen_direct = false; // HERE: control flag to remember if we saw the text node
        loop {
            let node = match deserializer.next() {
                Some(result) => result?,
                None => break,
            };
            match node {
                Node::Attribute(attr) => {
                    let id = deserializer.attribute_id(&attr)?;
                    let field = __Attributes::__Ignore;
                    match field {
                        __Attributes::__Ignore => {}
                    }
                }
                Node::Open(data) => {
                    let id = deserializer.element_id(&data)?;
                    let element = __Elements::__Ignore;
                    match element {
                        __Elements::__Ignore => {
                            let mut nested = deserializer.nested(data);
                            nested.ignore()?;
                        }
                    }
                }
                Node::Text(text) => {
                    seen_direct = true; // HERE: set control flag
                    let mut nested = deserializer.for_node(Node::Text(text));
                    <String>::deserialize(&mut __value0, "ArtUri::uri", &mut nested)?;
                }
                node => {
                    return Err(
                        Error::UnexpectedNode({
                            let res = ::alloc::fmt::format(
                                format_args!("{0:?} in {1}", node, "ArtUri"),
                            );
                            res
                        }),
                    );
                }
            }
        }
        // HERE: this is the #after_loop code, same as the Node::Text code but with an
        // implied empty string value
        if !seen_direct {
            let mut nested = deserializer.for_node(Node::Text("".into()));
            <String>::deserialize(&mut __value0, "ArtUri::uri", &mut nested)?;
        }
        *into = Some(Self {
            uri: __value0.try_done("ArtUri::uri")?,
        });
        Ok(())
    }
```

File: instant-xml/src/impls.rs [hunk 1]
```diff
@@ -270,9 +270,9 @@ impl<'xml> FromXml<'xml> for String {
         }
 
         match deserializer.take_str()? {
-            Some(value) => *into = Some(value.into_owned()),
-            None => return Ok(()),
-        }
+            Some(value) => into.replace(value.into_owned()),
+            None => into.replace(String::new()),
```
Comment (djc) [inline instant-xml/src/impls.rs:274]: You've fixed the `String` impl here -- do we need similar fixes for `&str` and `Cow<'_, str>`?

File: instant-xml-macros/src/de.rs [hunk 2]
```diff
@@ -624,10 +642,21 @@ fn named_field<'a>(
         } else if field_meta.direct {
             direct.extend(quote!(
                 Node::Text(text) => {
+                    seen_direct = true;
```
Comment (djc) [inline instant-xml-macros/src/de.rs:645]: Does this result in different handling between self-closing tags and empty text nodes? Is that desirable?

File: instant-xml/src/impls.rs (hunk 1 above)
Comment (wez) [reply instant-xml/src/impls.rs]: Yeah; I've added `Cow<'_, str>` in the latest push. There is no impl for `str`.

File: instant-xml-macros/src/de.rs (hunk 2 above)
Comment (wez) [reply instant-xml-macros/src/de.rs:645]: The behavior is the same for both; I've added a couple of test cases to demonstrate that.
I think having them behave consistently with each other is reasonable.
I don't need to make a distinction in my downstream code.

I imagine that if you did care about the difference, you would want the direct value to be `Option<String>`:

```rust
#[derive(Debug, PartialEq, FromXml)]
struct OptionArtUri {
    #[xml(direct)]
    uri: Option<String>,
}
```

with the current state in this PR, we'll produce `Ok(OptionContainer { art: Some(OptionArtUri { uri: Some("") }) })` for that case.

I'm not sure if there's a reasonable way to make that come back as `uri: None` without making more significant changes to the traits and types, and/or having some understanding about whether the target type is `Option`.

Without this PR, this case produces a `MissingValue` error, so I don't think that things are worse off than they were.

Comment (djc) [reply instant-xml-macros/src/de.rs:645]: Agree that this behavior is desired.

**Comment by djc** [comment]
This makes sense to me! I rebased and made some minor style changes in the `impls` module.

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
* Published instant-xml v0.5.0 at registry `crates-io`
* [new tag]         0.5.0 -> 0.5.0

**Pull request merged**